src/app/editor/contextView.js

```diff
@@ -129,11 +130,36 @@ class ContextView {
       }
     }
 
+    function next () {
+      var currentName = node.attributes.name
+      if (currentName === this.refName) {
+        this.ref = this.ref === undefined ? 0 : this.ref
+      } else { this.ref = 0 }
+      var nodes = self._api.contextualListener._activeHighlights
+      self._api.jumpTo(nodes[this.ref].position)
+      this.ref = (this.ref + 1) % nodes.length
+      this.refName = currentName
+    }
+
+    function previous () {
+      var currentName = node.attributes.name
+      if (currentName === this.refName) {
```

Comment:
im not really sure you need this comparison...
cause anyway, `this.ref` can be reset to `0` in  `_render` ...

Reply:
I need to reset it each time **new keyword** is chosen. If I reset it in the _render it will reset to 0 each time you click next/previous. So, any ideas how else to reset it each time we choose new keyword?

Reply:
@yann300 One more problem. The search is not working as it should. With 'getActiveHighlights' I don't get all the highlighted keywords (references + declaration) - sometimes there are all, sometimes not.

I think the key is in this function, I will try to figure it out myself, but maybe you can help me understand what is happening in this function

![screen shot 2017-11-28 at 6 16 17 am](https://user-images.githubusercontent.com/1076427/33303647-a39e10f0-d404-11e7-9c06-c9b339483f24.png)

Here's also my logs from last commit - they show all nodes when I click next and the array of nodes is changing so the PR is not finished yet.
![screen shot 2017-11-28 at 6 23 56 am](https://user-images.githubusercontent.com/1076427/33303674-df330986-d404-11e7-8310-f9f30adbd235.png)

Reply:
the search? do you mean the `previous/next` ?

Reply:
If there are 4 references, we are looping over 5 elements (the declaration is included). I think this is why you miss one.
but I think ok fine like that.
Note that there's a misbehavior somewhere the `up arrow` does not work as expected...

Reply:
Check my logs above. I am logging all nodes (their start position) with each next click. And as you can see I get 5 nodes, but not always...

 
![screen shot 2017-11-28 at 4 03 06 pm](https://user-images.githubusercontent.com/1076427/33326795-343c3e06-d456-11e7-8edb-51add086e358.png)

Reply:
let 's say that you have 4 ref and 1 declaration.
If the cursor is on a declaration, we do not highlight it, so you'll get 4 `_activeHighlights`
If the cursor is on a ref, we do highlight the declaration, so you'll get 5 `_activeHighlights` .
we could change this code to:
```
if (node.attributes && node.attributes.referencedDeclaration) {
      highlights(node.attributes.referencedDeclaration)
      var current = this._index['FlatReferences'][node.attributes.referencedDeclaration]
      this._highlight(current, compilationResult)
    } else {
      highlights(node.id)
      this._highlight(node, compilationResult)
    }
```

Reply:
to always highlight the declaration. not sure it is ui wise but let see (note that most IDE highlight the declaration as well - but only the name, not the entire line ...).
We should improve at some point the highlighter to highlight the name and not the entire line...